### DISCUSSION TITLE
Package manager

### DISCUSSION THREAD

**Discussion by pmoura** [LogtalkDotOrg/logtalk3#118] [Ideas]
Work on a Logtalk package manager is in progress. This discussion is the preferred medium for discussing its goals and features and providing feedback. The current expectation is to have a beta version included in the upcoming 3.51.0 release.

**Documentation preview**

https://logtalk3.readthedocs.io/en/latest/devtools/packs.html

**Main goals**

- Portability: support for all Prolog backends
- Decentralized: support for multiple registries (e.g. official, public, company private, ...)
- Secure: checksums and signed packs, sanitized registry and pack specs

**Main concepts**

- Pack: a Logtalk (or Prolog) application or library
- Registry: a collection of pack specifications

**Main features**

- Add, pin, update, and delete registries
- Install, pin, update, and uninstall packs
- List available, installed, outdated, orphaned, and dependent packs
- Easy to express pack dependencies (names and versions)
- Support for git repos, archives, and local directories for registries
- Local and remote registry and pack specifications
- Support for specifying multiple pack versions
- Minimal specification of registries and packs (using Logtalk objects implementing pack and registry protocols)

**Materialization**

A Logtalk `packs` tool, included in the Logtalk distribution.

**Timeline**

1. Stable registry and pack specification protocols (enabling earlier user experimentation and feedback)
2. Stable implementation of registry handling
3. Basic pack handling functionality
4. Dependencies handing

**Examples**

- Registry specification (details subject to change):

```logtalk
:- object(logtalk_packs_registry,
	implements(registry_protocol)).

	:- info([
		version is 1:0:0,
		author is 'Paulo Moura',
		date is 2021-02-10,
		comment is 'Official Logtalk packs registry.'
	]).

	name('logtalk-packs').

	description('Official Logtalk packs registry').

	home('https://github.com/LogtalkDotOrg/logtalk-packs').

	clone('https://github.com/LogtalkDotOrg/logtalk-packs.git').

	archive('https://github.com/LogtalkDotOrg/logtalk-packs/archive/main.zip').

:- end_object.
```

- Pack specification (details subject to change):

```logtalk
:- object(lflat_pack,
	implements(pack_protocol)).

	:- info([
		version is 1:0:0,
		author is 'Paulo Moura',
		date is 2021-02-07,
		comment is 'Pack manifest file for L-FLAT - Logtalk Formal Language and Automata Toolkit.'
	]).

	name(lflat).

	description('L-FLAT - Logtalk Formal Language and Automata Toolkit').

	license('MIT').

	home('https://github.com/l-flat/lflat').

	version(
		2:0:2,
		stable,
		'https://github.com/l-flat/lflat/archive/refs/tags/v2.0.2.tar.gz',
		sha256 - '8774b3863efc03bb6c284935885dcf34f69f115656d2496a33a446b6199f3e19',
		[logtalk >= 3:36:0],
		all
	).

:- end_object.
```

**Comment by hnbeck** [comment]
Hi, 

sounds good. This is the right context to ask something I want ask a long time. How to arrange source code such that packages are supported best. To be more concrete:

My app has code for handling OpenDrive files, writing xgml Graph files, communication. It is all part of my code, my one application. Now, how should I structure the code such that OpeDrive, xgml parts etc. could be coupled out later as an own package (or library) usable by others? And this question is even more valid in the context of the package manager to come :)

Cheers

Hans

**Reply by pmoura** [reply]
Do all those components of your application have their own loader file and tester/tests files? Does you application uses a master loader file that loads the loader files for each of the components?

P.S. You can use (if not already) Logtalk's `code_metrics` tool to access some relevant application metrics like _efferent coupling_, _afferent coupling_, and _instability_. The `diagrams` tool may also enable a quick overview on how independent or coupled are the application components.

**Reply by hnbeck** [reply]
Yes, exactly. At the moment, one class per file. Thanks for the hint related to Logtalk code_metrics tool

Cherrs

Hans

**Reply by hnbeck** [reply]
Hi Paulo, 

regarding to the new package manger, does it require a special layout of a GitHub repository, for example main loader file and package file in root dir or whatever ? (BTW Its on my todo list to read the new docs in detail :)

**Reply by pmoura** [reply]
The following section in the documentation preview (which reflects the current Logtalk git version) includes a directory layout example:

https://logtalk3.readthedocs.io/en/latest/devtools/packs.html#registry-specification

Btw, the current Logtalk git version includes an improved version of the `packs` tool. The plan is to release a new Logtalk stable version later this month with the improved tool (its API is stable but test coverage is a WIP).

**Reply by pmoura** [reply]
P.S. The tool supports `registries::lint/0` and `packs::lint/0` predicates so that you can check if your registry and pack definitions are up-to-specs.

**Reply by hnbeck** [reply]
Super, thanks :)

Cheers

Hans

**Comment by pmoura** [comment]
Implemented one of the safety measures today. Registry loader files and registry/pack specification files and objects are filtered using Logtalk's [term-expansion mechanism](https://logtalk.org/manuals/userman/expansion.html) to ensure that only the expected terms are actually loaded and no unexpected directive goals are called. If you're curious, the current definitions of the hook objects are [here](https://github.com/LogtalkDotOrg/logtalk3/blob/packs/tools/packs/safety_hooks.lgt).

**Reply by pmoura** [reply]
Also got basic functionality (adding a registry, installing a pack, ...) working on Windows 10 this weekend.

**Comment by pmoura** [comment]
Good progress implementing the package manager. Implemented functionality includes:

- add, update, clean, delete, and query registries
- install, pin, update, clean, delete, and query packs

Currently missing functionality includes:

- pack dependencies handling

Current version supports Linux, macOS, and Windows. But limited testing so far. To try out the current version, checkout the `packs` branch of the `logtalk3` repo and start with the documentation at `tools/packs/NOTES.md`. 

Stable pack registries (more to come):

https://github.com/LogtalkDotOrg/talkshow

Feedback most welcome.

**Reply by pmoura** [reply]
P.S. As the `packs` branch is rebased regularly, you will need to update it in your local clone using:
```bash
$ git checkout packs && git fetch && git reset --hard origin/packs
```

**Reply by pmoura** [reply]
Predicate options (verbose, force, clean, ...) and checking of pack archive signatures are now also implemented.

**Reply by pmoura** [reply]
Merged the `packs` branch into `master` as the tool is now sufficiently stable and usable. All future development will now take place in the `master` branch.

**Reply by pmoura** [reply]
A preview of the documentation of the new Logtalk `packs` tool (a portable package manager) is now available online at https://logtalk3.readthedocs.io/en/latest/devtools/packs.html Note that the link to the API documentation is not yet operational (as it points to the Logtalk website, which only includes the stable release documentation). Feedback most welcome, specially on the [security considerations](https://logtalk3.readthedocs.io/en/latest/devtools/packs.html#security-considerations).

**Reply by pmoura** [reply]
File URLs for both registries and packs implemented. This allows using local directories for registry and pack sources, simplifying development before pushing changes to a remote/shared location.

**Reply by pmoura** [reply]
Basic pack dependencies handling implemented.

**Reply by infradig** [reply]
It seems you need to add **.git** to the end of the registry pack name.

**Reply by pmoura** [reply]
Indeed the registry URL must end with either a ​.git​  or an archive extension (except when using a `file://` URL for a local directory). That's in the API documentation for the predicate ​`add/3​` predicate but not highlighted in the ​packs​ tool overview. Pushed documentation updates.

**Reply by pmoura** [reply]
Added a `registries::add/1` predicate to simplify the most common cases (for registry users, a registry made available as a git repo; for registry developers, a registry made available as a local directory).

**Reply by pmoura** [reply]
A list of publicly available Logtalk and Prolog pack registries is available at:

https://github.com/LogtalkDotOrg/pack-registries

**Reply by pmoura** [reply]
All [pack dependency](https://logtalk3.readthedocs.io/en/latest/devtools/packs.html#pack-dependencies) operators (which use all the standard term comparison operators) are now implemented in the current git version. Support for _range_ dependencies (e.g. `[reg::foo @>= 2.1, reg::foo < 3]`) is also implemented.

**Reply by pmoura** [reply]
Basic support for defining, saving, and restoring _virtual environments_ is now implemented. An overview of the new functionality is available at:
https://logtalk3.readthedocs.io/en/latest/devtools/packs.html#virtual-environments
